### PR TITLE
[Ray] Implement ray task executor progress

### PR DESCRIPTION
**Pull request by zhongchun**
## What do these changes do?
This pr implement the ray task executor progress function. `TaskProcessor` can get task progress by calling `RayTaskExecutor::get_progress`.

## Related issue number

#2893 

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

**Comment by fyrestone** [comment]
Can you enable some progress tests for Ray backend?

**Comment by wjsi** [comment]
No tests?

**Comment by zhongchun** [comment]
> Can you enable some progress tests for Ray backend?

Yes, I'll add some test.

**Comment by zhongchun** [comment]
> No tests?

I'll add some test. Thanks for you reminder.

File: mars/services/task/execution/ray/executor.py [hunk 1]
```diff
@@ -224,6 +230,10 @@ async def execute_subtask_graph(
                 # TODO(fyrestone): Fetch(not get) meta object here.
                 output_meta_object_refs.append(meta_object_ref)
             context.update(zip(output_keys, output_object_refs))
+        self._stage_output_object_refs.append(stage_object_refs)
+        prev_progress = sum(self._stage_tile_progresses)
+        curr_tile_progress = self._tile_context.get_all_progress() - prev_progress
+        self._stage_tile_progresses.append(curr_tile_progress)
```
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:236]: Is the tile progress calculation should be at the beginning of the `execute_subtask_graph`?  The progress will be incorrect if `get_progress` while submitting the subtask graph in current implementation.

Comment (fyrestone) [reply mars/services/task/execution/ray/executor.py:236]: Or can we calculate the tile progress when `get_progress` called?

File: mars/services/task/execution/ray/executor.py [hunk 2]
```diff
@@ -293,7 +303,27 @@ async def get_available_band_resources(self) -> Dict[BandType, Resource]:
 
     async def get_progress(self) -> float:
         """Get the execution progress."""
-        return 1
+        executor_progress = 0.0
+        assert len(self._stage_tile_progresses) == len(self._stage_output_object_refs)
```
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:307]: Any necessary to track all the output object refs of all stages?

File: mars/services/task/execution/ray/executor.py (hunk 1 above)
Comment (zhongchun) [reply mars/services/task/execution/ray/executor.py:236]: Yes, we should calculate tile progress before executing subtask graph and it's better than calculating every time in `get_progress`. Thanks for your reminder.

File: mars/services/task/execution/ray/executor.py (hunk 2 above)
Comment (zhongchun) [reply mars/services/task/execution/ray/executor.py:307]: Not necessary. Just calculating current stage progress is ok. I'll optimize this.

File: mars/services/task/execution/ray/executor.py [hunk 3]
```diff
@@ -249,6 +261,8 @@ async def execute_subtask_graph(
 
         logger.info("Waiting for stage %s complete.", stage_id)
         ray.wait(output_object_refs, fetch_local=False)
+        self._pre_all_stage_progress += self._calculate_cur_stage_progress()
```
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:264]: Here already waited the stage output object refs, can you provide a fast path?

File: mars/services/task/execution/ray/executor.py [hunk 4]
```diff
@@ -131,6 +131,12 @@ def __init__(
         self._task_context = {}
         self._available_band_resources = None
 
+        # For progress
+        self._pre_all_stage_progress = 0.0
```
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:135]:
```suggestion
        self._pre_all_stages_progress = 0.0
```

File: mars/services/task/execution/ray/executor.py [hunk 5]
```diff
@@ -131,6 +131,12 @@ def __init__(
...
+        self._pre_all_stage_tile_progress = 0
```
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:136]:
```suggestion
        self._pre_all_stages_tile_progress = 0
```

File: mars/services/task/execution/ray/executor.py (hunk 3 above)
Comment (zhongchun) [reply mars/services/task/execution/ray/executor.py:264]: We indeed to update `_pre_all_stage_progress ` after all output object refs are finished. It maybe right here.

Comment (zhongchun) [reply mars/services/task/execution/ray/executor.py]: We can just use `.self._cur_stage_tile_progress` as current stage progress instead of calculation.

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**